### PR TITLE
Issue/15632 moderate comments in reader posts

### LINKED ISSUE #15632: Allow moderation of threaded comments for users with appropriate rights

### PR DESCRIPTION
**Pull request by khaykov**
Fixes #15632

This PR adds ability to moderate comments in Reader posts for Admin and Editor users of the site that the post belongs to.

Comment moderation is performed after the Snackbar is dismissed. Because of the implementation of comment list, if you moderate a comment and immediately rotate the device, the comment will briefly reapear in the list. I worked on workaround, but the logic gets too complicated for such a rare usecase.

[Image from Gyazo](https://gyazo.com/9a70ef5199e7d1432681c847581f3fef)

To test:

- Confirm that the moderation menu is visible only to the users with Administrator or Editor roles on the site the post belongs to.
- Confirm that comment is properly moderated.
- Confirm that the undo action works.
- Confirm that the empty view is visible after last comment was removed from the list.
- Confirm that share still works.

## Regression Notes
1. Potential unintended areas of impact
- Moderation menu is visible to users with incorrect roles.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
- Manual testing.

3. What automated tests I added (or what prevented me from doing so)
- Legacy code is very unfriendly to testing.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/reader/utils/ReaderCommentLeveler.java [hunk 1]
```diff
@@ -36,10 +36,14 @@ public ReaderCommentList createLevelList() {
...
-            if (comment.level == 0 && comment.parentId != 0) {
-                comment.level = 1;
+        // a non-zero level to distinguish them from top level comments
+        ReaderCommentList remainingComments = new ReaderCommentList();
+        remainingComments.addAll(mComments);
+        remainingComments.removeAll(result);
+
+        for (ReaderComment comment : remainingComments) {
+            if (!hasParent(comment)) {
+                comment.level = 0;
```
Comment (develric) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/utils/ReaderCommentLeveler.java]: Just to check if I got it right, this means that we are considering/representing orphan comments as root comments, correct?

File: WordPress/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -414,6 +414,7 @@
     <string name="comment_reply_to_user">Reply to %s</string>
     <string name="comment_trashed">Comment trashed</string>
     <string name="comment_spammed">Comment marked as spam</string>
+    <string name="comment_unarppoved">Comment unapproved</string>
```
Comment (develric) [inline WordPress/src/main/res/values/strings.xml]: np: small typo in name "comment_unarppoved" vs "comment_unapproved"

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java [hunk 3]
```diff
@@ -656,7 +725,7 @@ private ReaderCommentAdapter getCommentAdapter() {
             // adapter calls this when user taps reply icon
             mCommentAdapter.setReplyListener(commentId -> setReplyToCommentId(commentId, true));
             // adapter calls this when user taps share icon
-            mCommentAdapter.setCommentShareListener(this::shareComment);
+            mCommentAdapter.setCommentShareListener(this::performCommentAction);
```
Comment (develric) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderCommentListActivity.java]: np: maybe we can consider renaming the `setCommentShareListener` to something like `setPerformCommentActionListener`; wdyt?

File: WordPress/src/main/java/org/wordpress/android/ui/reader/adapters/ReaderCommentMenuActionAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            @DrawableRes val iconRes: Int
+        ) : ReaderCommentMenuItem()
+
+        data class Divider(
+            override val type: ReaderCommentMenuActionType = DIVIDER_NO_ACTION
+        ) : ReaderCommentMenuItem()
+    }
+
+    enum class ReaderCommentMenuActionType {
+        APPROVE, UNAPROVE, SPAM, TRASH, EDIT, SHARE, DIVIDER_NO_ACTION
```
Comment (develric) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/adapters/ReaderCommentMenuActionAdapter.kt]: np: small typo in UNAPROVE vs UNAPPROVE

**Review by develric — COMMENTED** [review]
Hey @khaykov 👋 , code LGTM I just left a couple of np for you to check 🙇 ....also did a first testing round with a .com site and it worked as described. I'll do more testing later and tomorrow, let me know and thanks! 🙇

File: WordPress/src/main/java/org/wordpress/android/ui/reader/utils/ReaderCommentLeveler.java (hunk 1 above)
Comment (khaykov) [reply WordPress/src/main/java/org/wordpress/android/ui/reader/utils/ReaderCommentLeveler.java]: Yep 👍

**Comment by khaykov** [comment]
Added a feature flag. It's on by default, so you can either toggle it remotely, or toggle the boolean [here](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/15632-moderate-comments-in-reader-posts/WordPress/src/main/java/org/wordpress/android/util/config/ReaderCommentsModerationFeatureConfig.kt#L11).

**Review by develric — APPROVED** [review]
Thanks for the changes @khaykov ! code LGTM and works as described.

**Pull request merged**